src/INode.java

```diff
@@ -229,9 +251,63 @@ private ASTValue evalDefineProperty(WordsEnvironment environment) {
 		throw new AssertionError("Not yet implemented");	
 	}
 
-	private ASTValue evalEquals(WordsEnvironment environment) {
-		// TODO
-		throw new AssertionError("Not yet implemented");	
+	private ASTValue evalEquals(WordsEnvironment environment) throws WordsRuntimeException {
+		ASTValue lhs = children.get(0).eval(environment);
+		ASTValue rhs = children.get(1).eval(environment);
+		
+		// The special type Nothing is equal only to Nothing
+		if ((lhs.type == ValueType.NOTHING && rhs.type != ValueType.NOTHING) || 
+			(lhs.type != ValueType.NOTHING && rhs.type == ValueType.NOTHING)) {
+			return new ASTValue(false);
+		}
+		
+		if (lhs.type == ValueType.NOTHING && rhs.type == ValueType.NOTHING) {
+			return new ASTValue(true);
+		}
+		
+		// Objects only equal another object and cannot be equal to non-objects
+		if ((lhs.type == ValueType.OBJ && rhs.type != ValueType.OBJ) || 
+			(lhs.type != ValueType.OBJ && rhs.type == ValueType.OBJ)) {
+			return new ASTValue(false);
+		}
+		
+		if ((lhs.type == ValueType.OBJ && rhs.type == ValueType.OBJ) &&
+			(lhs.objValue == rhs.objValue)) {
+			return new ASTValue(true);
+		}
+		
+		// Remaining types must be a number or string
+		assert lhs.type == ValueType.NUM || lhs.type == ValueType.STRING : "Left side has type " + lhs.type.toString();
+		assert rhs.type == ValueType.NUM || rhs.type == ValueType.STRING : "Right side has type " + rhs.type.toString();
+		
+		// Number/string type coercion
+		if (lhs.type != rhs.type) {
+			if (lhs.type == ValueType.STRING) {
+				lhs = lhs.tryCoerceTo(ValueType.NUM);
+				
+				// If string to number coercion failed on lhs, do number to string coercion on rhs
+				if (lhs.type == ValueType.STRING)
+					rhs = rhs.tryCoerceTo(ValueType.STRING);
+			} else {
+				// lhs must be a number
+				rhs = rhs.tryCoerceTo(ValueType.NUM);
+				
+				// If string to number coercion failed on rhs, do number to string coercion on lhs
+				if (rhs.type == ValueType.STRING)
+					lhs = lhs.tryCoerceTo(ValueType.STRING);
+			}
+		}
+		
+		// lhs and rhs are now the same type
+		assert lhs.type == rhs.type : "lhs has type " + lhs.type.toString() + " and rhs has type " + rhs.type.toString();
+		switch (lhs.type) {
+			case NUM:
+				return new ASTValue(lhs.numValue == rhs.numValue);
```

Comment:
I think that, for a beginner language without integer representation, this is a good feature to have. It's an area where I would say Java is not intuitive, and we can choose to be.